### DISCUSSION TITLE
alternative syntax for `unapply` extractors

### DISCUSSION THREAD

**Discussion by erikerlandson** [scala/scala3#12793] [General Discussion]
pattern matching with `unapply` extractors requires the creation of a containing object, for example:
```scala
    object namedDU:
        def unapply(using Quotes)(u: quotes.reflect.TypeRepr): Option[Expr[NamedUnit]] =
            import quotes.reflect.*
            Implicits.search(TypeRepr.of[DerivedUnit].appliedTo(List(u, TypeBounds.empty))) match
                case iss: ImplicitSearchSuccess => Some(iss.tree.asExpr.asInstanceOf[Expr[NamedUnit]])
                case _ => None
```

I have been wondering if it would be more direct to replace `unapply` with an explicit name, but instead annotate it with something like `def extractor ...`, like:
```scala
def extractor namedDU(using Quotes)(u: quotes.reflect.TypeRepr): Option[Expr[NamedUnit]] =
    import quotes.reflect.*
    Implicits.search(TypeRepr.of[DerivedUnit].appliedTo(List(u, TypeBounds.empty))) match
        case iss: ImplicitSearchSuccess => Some(iss.tree.asExpr.asInstanceOf[Expr[NamedUnit]])
        case _ => None
```

Adding new keywords is not cost free but would reduce boilerplate

**Comment by erikerlandson** [comment]
more accurately,  `def extractor name ...` would be syntactic sugar for
```scala
object name:
    def unapply ...
```

**Comment by som-snytt** [comment]
Is this different from "name-based extractors"?
https://dotty.epfl.ch/docs/reference/changed-features/pattern-matching.html

Edit: I see they call it "Optionless". (No hyphen required for option-less.) "De-optioned."

**Reply by erikerlandson** [reply]
It is different: If I want to create an extractor named `foo`, the current syntax requires me to create an `object foo`

```scala
object foo:
    def unapply ...
```

IMO it seems a bit less boilerplate to declare the extractor directly

```scala
def extractor foo ...
```
or even

```scala
extractor foo ... 
```

**Reply by erikerlandson** [reply]
In other words, the `object` really only exists to give the extractor a unique name, which seems inelegant

**Reply by erikerlandson** [reply]
clearly one doesn't want to change the existing system, just proposing a new syntactic sugar for the existing system that captures the semantics a bit more clearly (in my opinion)

**Reply by erikerlandson** [reply]
yet another way to express it: "extractor" is really a first-class concept in scala, and quite powerful, which doesn't in principal need to be tied to an "object" - it is an entity in its own right, and I propose it merits its own keyword `extractor`.  If it remains implemented as `unapply` underneath an `object` of a given name, that is just a smart design strategy

**Reply by som-snytt** [reply]
Thanks, I did have coffee this morning, didn't I? When I tried it, below, I see what you mean.

**Comment by nafg** [comment]
There could be a SAM that would let you create one with an ordinary function

**Reply by erikerlandson** [reply]
What is `SAM` ?

**Reply by som-snytt** [reply]
"single abstract method".  `Extractor[A, B]` would take an A.

**Comment by som-snytt** [comment]
I see the explicit object is required for extension method version, otherwise "constructor proxy object C cannot be used as a value". It would help to lift that restriction for this context.
```
class C(val i: Int):
  override val toString = s"C-of($i)"
object C
/*
object C:
  def unapply(x: Any): Option[C] =
    x match
    case y: C if y.i > 42 => Some(y)
    case _ => None
*/

extension (um: C.type) def unapply(x: Any): Option[C] =
  x match
  case y: C if y.i > 42 => Some(y)
  case _ => None

@main def test() = println {
  val z: Any = new C(55)
  z match {
    case C(c) => c
  }
}
```

**Comment by nafg** [comment]
Single Abstract Method.


trait Extractor[-A, +B] { def unapply(in: A): Option[B] }

val AsInt: Extractor[String, Int] = _.toIntOption

"10" match { case AsInt(i) => i - 1 }

**Reply by erikerlandson** [reply]
would that allow the more complex curried parameter signatures, such in my original example at the top?
```scala
def unapply(using Quotes)(u: quotes.reflect.TypeRepr): Option[Expr[NamedUnit]]
```

**Comment by nafg** [comment]
No.

Of course you could make AsInt a def that takes implicit parameters to
create the extractor.

**Comment by erikerlandson** [comment]
Here's yet a third idea: in any `match` statement:
```scala
// a is of type A, b is of type B
match a
    case foo(b) => ...
```
the compiler simply admits any function `foo` if it is of type `A => Option[B]` as an extractor.
Likewise for all the usual extractor varations `A=>Boolean`, `A=>Option[(B, C, ...)]`, etc

**Reply by erikerlandson** [reply]
I say "simply" without knowing how simple it really is, but clearly it is already checking the signatures of any `unapply` extractors, and so extending that to checking the corresponding signatures of functions like `foo` seems superficially the same

**Reply by erikerlandson** [reply]
This seems like the right answer, the more I think about it. No new keywords or privileged types. The compiler allows any function/method of the proper signature to behave as an extractor when it is being used in such a context.

**Reply by erikerlandson** [reply]
https://github.com/lampepfl/dotty-feature-requests/issues/218
